src/3d/shaders/mesh/mesh.frag

```diff
@@ -0,0 +1,275 @@
+#version 330 core
+
+in treatedVertex {
+    vec3 wordPosition;
+    vec3 wordNormal;
+    float magnitude;
+    noperspective vec4 edgeA;
+    noperspective vec4 edgeB;
+    flat int configuration;
+} fs_in;
+
+out vec4 fragColor;
+
+uniform bool wireframeEnabled;
+uniform float lineWidth;
+uniform vec4 lineColor;
+
+uniform int textureType;
+uniform int colorRampType;
+uniform vec4 meshColor;
+uniform float verticaleScale;
+
+uniform sampler1D colorRampTexture;
+uniform int colorRampCount;
+
+const int MAX_LIGHTS = 8;
+const int TYPE_POINT = 0;
+const int TYPE_DIRECTIONAL = 1;
+const int TYPE_SPOT = 2;
+struct Light {
+    int type;
+    vec3 position;
+    vec3 color;
+    float intensity;
+    vec3 direction;
+    float constantAttenuation;
+    float linearAttenuation;
+    float quadraticAttenuation;
+    float cutOffAngle;
+};
+uniform Light lights[MAX_LIGHTS];
+uniform int lightCount;
+
+
+vec4 wireframeShadeLine( const in vec4 color )
+{
+    // Find the smallest distance between the fragment and a triangle edge
+    float d;
+    if ( fs_in.configuration == 0 )
+    {
+        // Common configuration
+        d = min( fs_in.edgeA.x, fs_in.edgeA.y );
+        d = min( d, fs_in.edgeA.z );
+    }
+    else
+    {
+        // Handle configuration where screen space projection breaks down
+        // Compute and compare the squared distances
+        vec2 AF = gl_FragCoord.xy - fs_in.edgeA.xy;
+        float sqAF = dot( AF, AF );
+        float AFcosA = dot( AF, fs_in.edgeA.zw );
+        d = abs( sqAF - AFcosA * AFcosA );
+
+        vec2 BF = gl_FragCoord.xy - fs_in.edgeB.xy;
+        float sqBF = dot( BF, BF );
+        float BFcosB = dot( BF, fs_in.edgeB.zw );
+        d = min( d, abs( sqBF - BFcosB * BFcosB ) );
+
+        // Only need to care about the 3rd edge for some configurations.
+        if ( fs_in.configuration == 1 || fs_in.configuration == 2 || fs_in.configuration == 4 )
+        {
+            float AFcosA0 = dot( AF, normalize( fs_in.edgeB.xy - fs_in.edgeA.xy ) );
+            d = min( d, abs( sqAF - AFcosA0 * AFcosA0 ) );
+        }
+
+        d = sqrt( d );
+    }
+
+    // Blend between line color and phong color
+    float mixVal;
+    if ( d < lineWidth - 1.0 )
+    {
+        mixVal = 1.0;
+    }
+    else if ( d > lineWidth + 1.0 )
+    {
+        mixVal = 0.0;
+    }
+    else
+    {
+        float x = d - ( lineWidth - 1.0 );
+        mixVal = exp2( -2.0 * ( x * x ) );
+    }
+
+    return mix( color, lineColor, mixVal );
+}
+
+void adModel(const in vec3 worldPos,
+             const in vec3 worldNormal,
+             out vec3 diffuseColor)
```

Comment:
Is it possible to reuse the existing adsModel implementation that exists in light.inc.frag? or is it not possible because of something? It is quite a lot of duplicated code...

Reply:
I will try, but i have duplicate code to have more simple shader (not the apparent one, but the compiled one) without other light fucntion and other stuff we don't need.
And I wonder, if the issue didn't come from that include...
I will try.

Reply:
Finaly, it is ok, so ok...